src/main/java/com/google/devtools/build/lib/runtime/AuthHeadersProvider.java

```diff
@@ -0,0 +1,48 @@
+package com.google.devtools.build.lib.runtime;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Generic interface to provide authentication headers for http/grpc requests
+ * to bazel modules.
+ *
+ * <p>Implementations need to be thread-safe.
+ */
+@ThreadSafe
+public interface AuthHeadersProvider {
+
+  /**
+   * Returns the type of authentication mechanism used i.e. oauth.
+   */
+  String getType();
+
+  /**
+   * Returns request headers necessary for authentication to be added to the http/grpc request.
+   *
+   * <p>For auth tokens with a TTL attached an implementation is expected to transparently
+   * refresh the tokens and to always return headers with sufficient TTL left to complete
+   * the request.
+   */
+  Map<String, List<String>> getRequestHeaders(URI uri) throws IOException;
```

Comment:
should this be a Multimap?

Reply:
I had that initially but the most commonly used implementation (google auth) returns a `Map<String, List<String>>` in its corresponding method and so I figured it's nicer to not have to do needless conversion (and allocation) on every RPC. I don't have a strong opinion on it though and would be happy to change it if you insist.